### PR TITLE
Allow multiple macvlan networks to share a parent

### LINKED ISSUE #47317: [macvlan] Same parent and gateway for multiple networks
### Description

I'm remaking this from https://github.com/moby/libnetwork/issues/2384

It's still relevant to me (esp with ipv6 networking) and I'd be happy to try working on it, but wanted to check first if the solution proposed there had any issues.

### PR DESCRIPTION
**Pull request by andrewbaxter**
- Closes #47317

The only case where macvlan interfaces are unable to share a parent is when the macvlan mode is passthru. This change tightens the check to that situation.

**- What I did**

1. If two macvlan networks have the same parent and id -> already exists
2. If two macvlan networks have the same parent, different ids, and at least one is passthru -> error
3. If two macvlan networks have the same parent, different ids -> create network
4. If two macvlan networks have different parents -> create network

**- How I did it**

A new if and continue to skip marking as "found existing".

**- How to verify it**

Create two macvlan networks with the same parent.

**- Description for the changelog**

**- Description for the changelog**

```markdown changelog
Allow multiple macvlan networks with the same parent [moby#47318](https://github.com/moby/moby/pull/47318).
```

**- A picture of a cute animal (not mandatory but encouraged)**

image

[Man sets fire to smoke out opossums, burns down house instead](http://www.cbsnews.com/news/man-tries-to-scare-off-opossums-by-setting-fire-burns-down-own-house-instead/)

### CODE REVIEW

**Comment by andrewbaxter** [comment]
Just a note, I haven't managed to get tests running locally.  I built this with nix, which I think _might_ run tests as part of the build.  I tested it manually with with two macvlan interfaces with the same parent (`eth0`) and it appears to be working.

**Comment by thaJeztah** [comment]
Thank you for contributing! It appears your commit message is missing a DCO sign-off,
causing the DCO check to fail.

We require all commit messages to have a `Signed-off-by` line with your name
and e-mail (see ["Sign your work"](https://github.com/moby/moby/blob/v20.10.11/CONTRIBUTING.md#sign-your-work)
in the CONTRIBUTING.md in this repository), which looks something like:

```
Signed-off-by: YourFirsName YourLastName <yourname@example.org>
```

There is no need to open a new pull request, but to fix this (and make CI pass),
you need to _amend_ the commit(s) in this pull request, and "force push" the amended
commit.

Unfortunately, it's not possible to do so through GitHub's web UI, so this needs
to be done through the git commandline.

You can find some instructions in the output of the DCO check (which can be found
in the "checks" tab on this pull request), as well as in the [Moby contributing guide](https://github.com/moby/moby/blob/v20.10.11/docs/contributing/set-up-git.md).

Steps to do so "roughly" come down to:

1. Set your name and e-mail in git's configuration:

    ```bash
    git config --global user.name "YourFirstName YourLastName"
    git config --global user.email "yourname@example.org"
    ```

    (Make sure to use your _real_ name (**not your GitHub username/handle**) and e-mail)


2. Clone your fork locally
3. Check out the branch associated with this pull request
4. Sign-off and amend the existing commit(s)

    ```bash
    git commit --amend --no-edit --signoff
    ```

    If your pull request contains multiple commits, either squash the commits (if
    needed) or sign-off each individual commit.

5. _Force push_ your branch to GitHub (using the `--force` or [`--force-with-lease`](https://stackoverflow.com/questions/52823692/git-push-force-with-lease-vs-force) flags) to update the pull request.


Sorry for the hassle (I wish GitHub would make this a bit easier to do), and let me know if you need help or more detailed instructions!

**Comment by thaJeztah** [comment]
Oh! If you're updating; could you remove the `Closes #47317` from the commit message? Unfortunately, GitHub has a tendency to either close related tickets prematurely (or producing "noice / notifications") when commits are merged in various places (which could be in forks of this repository). For that reason, we try to avoid referencing tickets in _commit_ messages (and only put them in the description on GitHub, not the commit message).

**Comment by andrewbaxter** [comment]
Oh sure, and sorry about the DCO thing! I should re-read contributing more carefully...

**Comment by thaJeztah** [comment]
No worries! iIt happens frequently (hence my "canned reply" above, which I wrote so that I didn't have to come up with a good way to describe the steps every time 😂 ❤️)

**Comment by thaJeztah** [comment]
Oh! Looks like there's no e-mail address in your sign-off; 

```
Signed-off-by: Andrew Baxter <>
```

File: libnetwork/drivers/macvlan/macvlan_network.go
Comment (robmry) [inline libnetwork/drivers/macvlan/macvlan_network.go:70]: The function `getDummyName()` is used to generate the name for a dummy link when no parent interface is specified - it just translates `foo` into `dm-foo`. So, for `nwName` - as the actual name doesn't seem to be stored by the driver, its short-id is just `stringid.TruncateID(nw.config.ID)`.

It might be worth making the error message simpler and more specific by splitting out the two cases...

```suggestion
                                if config.MacvlanMode == modePassthru {
                                        return false, fmt.Errorf(
                                                "cannot use mode passthru, macvlan network %s is already using parent interface %s",
                                                stringid.TruncateID(nw.config.ID),
                                                config.Parent,
                                        )
                                } else if nw.config.MacvlanMode == modePassthru {
                                        return false, fmt.Errorf(
                                                "macvlan network %s is already using parent interface %s in mode passthru",
                                                stringid.TruncateID(nw.config.ID),
                                                config.Parent,
                                        )
                                }
                                continue
```

**Review by robmry — CHANGES_REQUESTED** [review]
Thank you for looking at this, it looks like the right approach - but there are a couple of issues ...

Integration test, `TestDockerNetworkMacvlan/OverlapParent` in [macvlan_test.go](https://github.com/moby/moby/blob/master/integration/network/macvlan/macvlan_test.go) was supposed to be checking it's not possible to create two macvlan networks with the same parent.

Unfortunately, the test hasn't been running for a while. But, when that's fixed (https://github.com/moby/moby/pull/47382), it'll fail with this change.

Also, with a sequence like this ...

```
# docker network create -d macvlan mv1
# docker network create -d macvlan -o parent=dm-8a604e8c9839 mv2
# docker network rm mv1
# docker run --rm -ti --network mv2 alpine
docker: Error response from daemon: the requested parent interface dm-8a604e8c9839 was not found on the Docker host.
```

... the first 'create' doesn't specify a parent interface, so a dummy link is created ("dm-8a604e8c9839"). That's remembered in `config.CreatedSlaveLink`, and the link is deleted by `driver.DeleteNetwork()` when the network's removed.

That was ok when the dummy link could only be the parent for one macvlan network but now, deleting `mv1` leaves `mv2` without a parent (and so, would break connectivity for containers on `mv2`).

Are you up for dealing with that and updating the integration test? I think we should also have a new tests for the pass-through special case, and the create/delete sequence above - if that's permitted.

**Comment by andrewbaxter** [comment]
Thanks for the review! And good point on the logs, I applied the suggestion.

> Thank you for looking at this, it looks like the right approach - but there are a couple of issues ...
> 
> Integration test, `TestDockerNetworkMacvlan/OverlapParent` in [macvlan_test.go](https://github.com/moby/moby/blob/master/integration/network/macvlan/macvlan_test.go) was supposed to be checking it's not possible to create two macvlan networks with the same parent.
> 
> Unfortunately, the test hasn't been running for a while. But, when that's fixed (#47382), it'll fail with this change.
> 

Roger, I'll try to update that.

> Also, with a sequence like this ...
> 
> ```
> # docker network create -d macvlan mv1
> # docker network create -d macvlan -o parent=dm-8a604e8c9839 mv2
> # docker network rm mv1
> # docker run --rm -ti --network mv2 alpine
> docker: Error response from daemon: the requested parent interface dm-8a604e8c9839 was not found on the Docker host.
> ```
> 
> ... the first 'create' doesn't specify a parent interface, so a dummy link is created ("dm-8a604e8c9839"). That's remembered in `config.CreatedSlaveLink`, and the link is deleted by `driver.DeleteNetwork()` when the network's removed.
> 
> That was ok when the dummy link could only be the parent for one macvlan network but now, deleting `mv1` leaves `mv2` without a parent (and so, would break connectivity for containers on `mv2`).

Off the top of my head, three ways to handle this would be:
1. Keep track of children and only delete the interface when no more children exist
2. Don't delete the dummy(/vlan) parent interface, but instead clean it up in system prune
3. Re-restrict this so that sharing parents is only allowed for existing interfaces

I think the 1/2 would be best - if it's technically allowed someone might have a legitimate use case, but I'm not sure if there's precedence for that sort of reference counting in the code base.

For 1 I guess iterating all the networks and counting parent users before deleting would probably be the simplest/safest. Reference counting, associating a set of children, deleting the parent in a finalizer are other options but have probably worse drawbacks.

For 2, if someone wanted to share dummy parent interfaces they might want it to keep existing even if they delete all the users. If they're really going to do that, would it be easier to do it outside of moby (i.e. create the interface with their own tooling, the same that would look up the dummy interface id)? If so maybe 3 would be better, otherwise for 2 maybe the dummy interface could be turned into a normal network that could be managed with network tools.

Sorry, I'm probably missing some context here so my guesses may be wildly off.

Did you have any thoughts on how this should be handled?

> 
> Are you up for dealing with that and updating the integration test? I think we should also have a new tests for the pass-through special case, and the create/delete sequence above - if that's permitted.

Right, makes sense.  I'll look into doing those too.

Thanks again for the review!

**Comment by robmry** [comment]
> 1. Keep track of children and only delete the interface when no more children exist
> 2. Don't delete the dummy(/vlan) parent interface, but instead clean it up in system prune
> 3. Re-restrict this so that sharing parents is only allowed for existing interfaces

Leaving it to system-prune might be tricky, we'd need to be certain that the interface is a docker-created dummy (just deleting interfaces named "dm-nnnn" when there's no "nnnn" network in-use might zap something it shouldn't). And, it'd rely on people running prune to declutter, when we could have dealt with when the last network went-away.

Disallowing sharing of a moby-created dummy interface (a variant on '3') would solve the problem, and this PR would still be a step forward in terms of being able to share real interfaces. It'd be easy to do by just raising an error if `nw.config.CreatedSlaveLink` is set on the existing interface, as another `else if` when checking for `passthru` networks. But, there might be a reason to want to share one of those interfaces for testing... although I'm not convinced and, as you say, it'd be possible to create the dummy parent outside of moby as a workaround.

So, some form of option 1 might be best ...

I agree that explicitly linking the networks or keeping a reference count could get messy.

Maybe set `config.CreatedSlaveLink` on the new network, if an existing network with the same parent already has the flag set. Then, everything sharing that moby-created dummy parent ends up with the flag set.

Then - in the delete, if the flag is set, search for another network with the same parent - only removing the dummy interface if none are found. That way, the search is only necessary if the flag is set, the order of deletions doesn't matter, and it's not necessary to work out whether the interface is a dummy created for some other already-deleted network.

> Right, makes sense. I'll look into doing those too.

Brilliant, thank you!

**Comment by andrewbaxter** [comment]
> So, some form of option 1 might be best ...

I was thinking about this a bunch and I think for vlan parents we'd need 1 either way, so probably I was focusing on the dummy interfaces too much.

I'll probably need some time for your updates to sink in to my brain too but that all makes sense.

I'll start by trying 1 iterating the networks and counting uses at delete time and see where that gets me.

**Comment by robmry** [comment]
> I was thinking about this a bunch and I think for vlan parents we'd need 1 either way, so probably I was focusing on the dummy interfaces too much.

Oh, yes - my description's also focussed on the dummy interfaces ... but it applies equally to the moby-created VLAN interfaces.

The suggestion is just to make sure `CreatedSlaveLink` is set on all the networks sharing a moby-created parent, not just the one that actually created it. Then it doesn't matter which order networks are deleted in, the parent interface for a network with `CreatedSlaveLink` set can be removed when no other networks have the same parent.

**Comment by thaJeztah** [comment]
DOH! Looks like the "accept suggestion" added a separate commit, which.. of course doesn't get a DCO sign-off.

It's probably fine to squash the commits into a single one (you can include the co-authored-by, but we're not always very fussy about that if it's suggestions from maintainers - it's usually fine to skip 😅)

**Comment by andrewbaxter** [comment]
> The suggestion is just to make sure `CreatedSlaveLink`

Right, that makes sense.

> It's probably fine to squash the commits into a single one

Ah sorry, I was intending to do that when I got back to this.  I'll do that now.

**Comment by andrewbaxter** [comment]
Okay, I think I've taken care of everything: counting other users, integration tests.

Notes on the use counting - I'm not sure if the `ok := ...; ok` is a convention, but I'd have to add another if/level of nesting to check the use count so I collapsed the 3 clauses into a single if. The counting could be delayed until it was confirmed to be a dummy/vlan parent, but the vlan code has the "is this a vlan device" checks within the delete/create functions and I'm not sure how hot this path is in real life so I went with the less disruptive change.

Notes on the tests - I think these were the key things to check:
- Sharing a dummy/vlan/physical/etc interface
- Deleting the original interface first
- Deleting the 2nd interface first
- Existing interfaces don't get deleted
- Sharing a parent with a passthru
- Creating a passthru with a shared parent

I didn't see explicit tests for ex: implicitly creating a dummy interface, and the existing overlap test would test one of the delete cases.  I'm not sure if it's better to have one test per thing to check or if because integration tests are slower it's better to combine them when reasonably possible, but I believe all of the new code paths are reasonably covered with the new test cases.

**Review by robmry — APPROVED** [review]
LGTM - thank you!

**Comment by thaJeztah** [comment]
@akerouanton PTAL

**Comment by akerouanton** [comment]
@andrewbaxter I'll take a look tomorrow.

**Comment by thaJeztah** [comment]
Tentatively adding this to the v27.0 milestone (assuming review doesn't bring up concerns).

I have a small follow-up to this PR (some cleaning up that I think could be done while I looked at the code); we can merge that one once this PR is approved/merged;

- https://github.com/moby/moby/pull/47994


❓ Does the same apply to ipvlan; do we need a follow-up to update that driver as well with the same changes?

File: integration/internal/network/ops.go
Comment (thaJeztah) [inline integration/internal/network/ops.go:49]: Looks like this no longer compiles when merged into master; I noticed this was failing when I restarted  Jenkins (which does a merge before testing);

[2024-06-15T13:02:18.893Z] # github.com/docker/docker/integration/internal/network
[2024-06-15T13:02:18.893Z] integration/internal/network/ops.go:49:47: undefined: types
[2024-06-15T13:02:18.893Z] integration/internal/network/ops.go:50:17: undefined: types


This needs to be updated to use `network.CreateOptions`.


Let me do a quick rebase, and fix this so that Ci can do another run 👍

**Comment by thaJeztah** [comment]
@andrewbaxter I did a quick rebase, and fixed that compile failure I mentioned above 😅 https://github.com/moby/moby/pull/47318#discussion_r1641325736

**Comment by andrewbaxter** [comment]
Oh thanks! Sorry, I should have done that.

I'm not sure if the ipvlan question was directed to me, my knowledge is pretty limited though. I'll try to read up on ipvlan in the meantime.

**Comment by thaJeztah** [comment]
> I'm not sure if the ipvlan question was directed to me, my knowledge is pretty limited though. I'll try to read up on ipvlan in the meantime.

Not necessarily; in all honesty, it's not my "forte" either, but I know the implementations for the ipvlan and macvlan drivers are very similar, so I was wondering if this was also a property they share (and if so, if we need to do a follow-up to make the changes there as well); for sure not a blocker for this pull request though!

**Review by corhere — APPROVED** [review]

**Review by akerouanton — APPROVED** [review]

**Comment by andrewbaxter** [comment]
Thanks again for helping me with this and the reviews!

**Pull request merged**
